<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · DinucleotideForces.jl</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://adigioacchino.github.io/DinucleotideForces.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>DinucleotideForces.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Dinucleotide-forces:-definition"><span>Dinucleotide forces: definition</span></a></li><li><a class="tocitem" href="#How-to-cite-this-package"><span>How to cite this package</span></a></li><li><a class="tocitem" href="#Examples"><span>Examples</span></a></li><li><a class="tocitem" href="#Known-issues"><span>Known issues</span></a></li><li><a class="tocitem" href="#API"><span>API</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/main/docs/src/index.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="DinucleotideForces"><a class="docs-heading-anchor" href="#DinucleotideForces">DinucleotideForces</a><a id="DinucleotideForces-1"></a><a class="docs-heading-anchor-permalink" href="#DinucleotideForces" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/adigioacchino/DinucleotideForces.jl">DinucleotideForces</a>, a Julia package to computed dinucleotide forces in nucleic-acid sequences.</p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>This package is not registered. Install with:</p><pre><code class="language-julia hljs">import Pkg
Pkg.add(url=&quot;https://github.com/adigioacchino/DinucleotideForces.jl&quot;)</code></pre><h2 id="Dinucleotide-forces:-definition"><a class="docs-heading-anchor" href="#Dinucleotide-forces:-definition">Dinucleotide forces: definition</a><a id="Dinucleotide-forces:-definition-1"></a><a class="docs-heading-anchor-permalink" href="#Dinucleotide-forces:-definition" title="Permalink"></a></h2><p>Nucleotide/dinucleotide forces for a sequence are parameters that capture pressures to increase or decrease the usage of a given nucleotide/dinucleotide in the sequence. So a positive force on a nucleotide corresponds to a pressure to have more occurrences of the given dinucleotide, while a negative pressure corresponds to a lack of occurrences of the dinucleotide in the considered sequence. Forces are inferred through a <a href="https://en.wikipedia.org/wiki/Principle_of_maximum_entropy">maximum entropy model</a>. For a more formal definition give a look at <a href="https://www.pnas.org/content/111/13/5054.short">this paper</a>.</p><h2 id="How-to-cite-this-package"><a class="docs-heading-anchor" href="#How-to-cite-this-package">How to cite this package</a><a id="How-to-cite-this-package-1"></a><a class="docs-heading-anchor-permalink" href="#How-to-cite-this-package" title="Permalink"></a></h2><p>If this package is used for research purposes, please consider citing <a href="https://doi.org/10.1093/molbev/msab036">this</a>  and <a href="https://www.pnas.org/content/111/13/5054.short">this</a> papers.</p><h2 id="Examples"><a class="docs-heading-anchor" href="#Examples">Examples</a><a id="Examples-1"></a><a class="docs-heading-anchor-permalink" href="#Examples" title="Permalink"></a></h2><h3 id="Basic-usage"><a class="docs-heading-anchor" href="#Basic-usage">Basic usage</a><a id="Basic-usage-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-usage" title="Permalink"></a></h3><p>We will take sub-sequence from the Influenza H5N1 PB2 segment (strain used: A/Anhui/1/2005) and we will use it as working example to test the function of the scripts. We start by loading the script and the sequence:</p><pre><code class="language-julia hljs">using DinucleotideForces
example_seq = &quot;ATGGAGAGAATAAAAGAATTAAGGGATCTAATGTCACAGTCCCGCACTCGCGAGATACTAACAAAAACCACTGTGGACCATATGGCCATAATCAAGAAGTACACATCAGGAAGACAAGAGAAGAACCCTGCTCTCAGAATGAAATGGATGATGGCAATGAAATATCCAATCACAGCGGACAAGAGAATAACAGAGATGATTCCTGAAAGGAATGAACAAGGGCAGACGCTCTGGAGCAAGACAAATGATGCCGGATCGGACAGGTTGATGGTGTCTCCCTTAGCTGTAACTTGGTGGAATAGGAATGGGCCGACGACAAGTGCAGTCCATTATCCAAAGGTTTACAAAACATACTTTGAGAAGGCT&quot;</code></pre><p>Let&#39;s start by computing the force on the CpG motif along the full genome: this can be done with the only exported function of the package, <code>DimerForce</code>, as shown in the following code block.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;CG&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 5 entries:
  &quot;A&quot;  =&gt; -1.06703
  &quot;T&quot;  =&gt; -1.73119
  &quot;C&quot;  =&gt; -1.547
  &quot;CG&quot; =&gt; -0.944914
  &quot;G&quot;  =&gt; -1.32423</code></pre><p>The resulting output will be a dictionary where the force is associated to each motif specified (in this case only the &quot;CG&quot; motif).</p><p>When no nucleotide biases (frequencies) are specified, as above, the corresponding forces are inferred together with the forces.  We will use the word <em>fields</em> to indicate forces on nucleotides, while <em>forces</em> will be from now on restricted to dinucleotides. Notice that <span>$\sum_n e^{h_n} = 1$</span>, where <span>$n \in \{A,C,G,T\}$</span>. This is the result of a gauge choice, because in general the model probabilities are invariant under the transformation $ h<em>n \to h</em>n+K$, which can then be used to make the fields interpretable as the logarithm of a frequency (more on this is discussed later in section <a href="#Focus-on-gauge-choices">Focus on gauge choices</a>).</p><p>A user-specified bias can be given, and in this case fields are not inferred:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; nt_bias = [0.25, 0.25, 0.25, 0.25]; # probs for A, C, G, T</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;CG&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs; freqs=nt_bias)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 1 entry:
  &quot;CG&quot; =&gt; -1.06519</code></pre><p>The package easily allows to compute forces on two or more dinucleotides:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;CG&quot;, &quot;TA&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 6 entries:
  &quot;A&quot;  =&gt; -1.0215
  &quot;T&quot;  =&gt; -1.58094
  &quot;C&quot;  =&gt; -1.64745
  &quot;CG&quot; =&gt; -0.864869
  &quot;G&quot;  =&gt; -1.42037
  &quot;TA&quot; =&gt; -0.754889</code></pre><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; nts = [&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [a*b for a in nts for b in nts];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 20 entries:
  &quot;T&quot;  =&gt; -1.76995
  &quot;C&quot;  =&gt; -1.53967
  &quot;CC&quot; =&gt; 0.0
  &quot;GC&quot; =&gt; 0.203128
  &quot;GG&quot; =&gt; 0.0
  &quot;CG&quot; =&gt; -1.25295
  &quot;AT&quot; =&gt; 0.0
  &quot;A&quot;  =&gt; -1.04555
  &quot;CA&quot; =&gt; 0.206446
  &quot;TG&quot; =&gt; 0.0859823
  &quot;TA&quot; =&gt; -0.144103
  &quot;GT&quot; =&gt; 0.0
  &quot;G&quot;  =&gt; -1.33292
  &quot;GA&quot; =&gt; 0.790941
  &quot;TT&quot; =&gt; 0.0
  &quot;AC&quot; =&gt; -0.185393
  &quot;CT&quot; =&gt; 0.0
  &quot;AA&quot; =&gt; 0.0
  &quot;AG&quot; =&gt; -0.639832
  &quot;TC&quot; =&gt; 0.403223</code></pre><p>Look at how the CpG force changed in the cases: this is due to the fact that different motifs interact. </p><h3 id="Focus-on-gauge-choices"><a class="docs-heading-anchor" href="#Focus-on-gauge-choices">Focus on gauge choices</a><a id="Focus-on-gauge-choices-1"></a><a class="docs-heading-anchor-permalink" href="#Focus-on-gauge-choices" title="Permalink"></a></h3><p>When the full set of fields and forces is inferred, the system of equations solved to obtain these parameters is underdetermined. This means that a gauge choice must be made, for instance by setting some forces to zero by passing less than 16 dinucleotides as motifs. If such a choise is not made, the package selects the gauge so that:</p><ul><li>the exponential of the fields sum to 1;</li><li>the forces for dinucleotides of the form NN and those of the form NT are put to zero.</li></ul><p>Notice that the script allows for flexible choices: if not all the dinucleotides are given, the gauge is always chosen such that the maximum possible number of the nucleotides of the form NN and NT have forces equal to zero, as one can check by running the next code blocks.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;AC&quot;, &quot;AG&quot;, &quot;AT&quot;, &quot;CA&quot;, &quot;GA&quot;, &quot;TA&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 10 entries:
  &quot;AC&quot; =&gt; -0.412762
  &quot;AT&quot; =&gt; 0.0
  &quot;A&quot;  =&gt; -0.999029
  &quot;T&quot;  =&gt; -1.53904
  &quot;C&quot;  =&gt; -1.67771
  &quot;AG&quot; =&gt; -0.373976
  &quot;CA&quot; =&gt; 0.563492
  &quot;G&quot;  =&gt; -1.46804
  &quot;TA&quot; =&gt; -0.365219
  &quot;GA&quot; =&gt; 0.652723</code></pre><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [
           &quot;AC&quot;,&quot;AG&quot;,&quot;AT&quot;,&quot;CA&quot;,
           &quot;CG&quot;,&quot;CT&quot;,&quot;GA&quot;,&quot;GC&quot;,
           &quot;GT&quot;,&quot;TA&quot;,&quot;TC&quot;,&quot;TG&quot;,
           ];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 16 entries:
  &quot;T&quot;  =&gt; -1.84946
  &quot;C&quot;  =&gt; -1.53652
  &quot;GC&quot; =&gt; 0.211443
  &quot;CG&quot; =&gt; -1.23487
  &quot;AT&quot; =&gt; 0.0
  &quot;A&quot;  =&gt; -1.0114
  &quot;CA&quot; =&gt; 0.175394
  &quot;TG&quot; =&gt; 0.211296
  &quot;G&quot;  =&gt; -1.33241
  &quot;TA&quot; =&gt; -0.0679201
  &quot;GT&quot; =&gt; 0.0
  &quot;GA&quot; =&gt; 0.75677
  &quot;AC&quot; =&gt; -0.178653
  &quot;CT&quot; =&gt; 0.0
  &quot;AG&quot; =&gt; -0.626449
  &quot;TC&quot; =&gt; 0.521892</code></pre><p>Notice that when dinucleotides are not given, this is equivalent to fix to 0 their forces.  Depending on which dinucleotides are not given, this might or not be a specific gauge - in other words, it might or not result in an equivalent model. For instance, the following three cells result in equivalent models (although the third has different parameters because it is in another gauge), while the fourth does not:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;AC&quot;, &quot;AG&quot;, &quot;AT&quot;, &quot;CA&quot;, &quot;GA&quot;, &quot;TA&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 10 entries:
  &quot;AC&quot; =&gt; -0.412762
  &quot;AT&quot; =&gt; 0.0
  &quot;A&quot;  =&gt; -0.999029
  &quot;T&quot;  =&gt; -1.53904
  &quot;C&quot;  =&gt; -1.67771
  &quot;AG&quot; =&gt; -0.373976
  &quot;CA&quot; =&gt; 0.563492
  &quot;G&quot;  =&gt; -1.46804
  &quot;TA&quot; =&gt; -0.365219
  &quot;GA&quot; =&gt; 0.652723</code></pre><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;AC&quot;, &quot;AG&quot;, &quot;CA&quot;, &quot;GA&quot;, &quot;TA&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 9 entries:
  &quot;AC&quot; =&gt; -0.412762
  &quot;A&quot;  =&gt; -0.999029
  &quot;T&quot;  =&gt; -1.53904
  &quot;C&quot;  =&gt; -1.67771
  &quot;AG&quot; =&gt; -0.373976
  &quot;CA&quot; =&gt; 0.563492
  &quot;G&quot;  =&gt; -1.46804
  &quot;TA&quot; =&gt; -0.365219
  &quot;GA&quot; =&gt; 0.652723</code></pre><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;AG&quot;, &quot;AT&quot;, &quot;CA&quot;, &quot;GA&quot;, &quot;TA&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 9 entries:
  &quot;AT&quot; =&gt; 0.537838
  &quot;A&quot;  =&gt; -0.998373
  &quot;T&quot;  =&gt; -1.56506
  &quot;C&quot;  =&gt; -1.65424
  &quot;AG&quot; =&gt; 0.097891
  &quot;CA&quot; =&gt; 0.0930896
  &quot;G&quot;  =&gt; -1.46444
  &quot;TA&quot; =&gt; -0.835621
  &quot;GA&quot; =&gt; 0.182321</code></pre><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;AG&quot;, &quot;CA&quot;, &quot;GA&quot;, &quot;TA&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 8 entries:
  &quot;A&quot;  =&gt; -1.00421
  &quot;T&quot;  =&gt; -1.46014
  &quot;C&quot;  =&gt; -1.76025
  &quot;AG&quot; =&gt; -0.174972
  &quot;CA&quot; =&gt; 0.365081
  &quot;G&quot;  =&gt; -1.47203
  &quot;TA&quot; =&gt; -0.563632
  &quot;GA&quot; =&gt; 0.454311</code></pre><p>In some cases (notably when all the dinucleotides are given), fixing fields is equivalent to fix (part of) a gauge, so the model is independendent on this choice (although the values of the inferred parameters depend on that).</p><h3 id="Inferring-forces-from-multiple-sequences"><a class="docs-heading-anchor" href="#Inferring-forces-from-multiple-sequences">Inferring forces from multiple sequences</a><a id="Inferring-forces-from-multiple-sequences-1"></a><a class="docs-heading-anchor-permalink" href="#Inferring-forces-from-multiple-sequences" title="Permalink"></a></h3><p>The package is able to infer fields and forces from a set of sequences:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;CG&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; seqs = [example_seq, example_seq];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(seqs, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 5 entries:
  &quot;A&quot;  =&gt; -1.06703
  &quot;T&quot;  =&gt; -1.73119
  &quot;C&quot;  =&gt; -1.547
  &quot;CG&quot; =&gt; -0.944914
  &quot;G&quot;  =&gt; -1.32423</code></pre><p>The sequences do not need to have the same length. However, the values of the inferred parameters weakly depend on the sequence length, that has to be provided</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;CG&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; seqs = [example_seq, example_seq[1:100]];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(seqs, motifs, 2500)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 5 entries:
  &quot;A&quot;  =&gt; -1.01189
  &quot;T&quot;  =&gt; -1.71435
  &quot;C&quot;  =&gt; -1.54984
  &quot;CG&quot; =&gt; -0.725426
  &quot;G&quot;  =&gt; -1.41015</code></pre><p>or it will arbitarily set by the package to the standard value of 5000</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [&quot;CG&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; seqs = [example_seq, example_seq[1:100]];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(seqs, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Warning: Lmodel is not specified and the sequence set provided contains sequences of different lengths. Using an arbitrary value of L=5000.
Dict{String, Float64} with 5 entries:
  &quot;A&quot;  =&gt; -1.01229
  &quot;T&quot;  =&gt; -1.71423
  &quot;C&quot;  =&gt; -1.54956
  &quot;CG&quot; =&gt; -0.726089
  &quot;G&quot;  =&gt; -1.40989</code></pre><p>Notice, in the last two examples, how the changes due to the different lengths used during the inference of the parameters only marginally affect their values.</p><h3 id="Computing-energies-and-loglikelihoods"><a class="docs-heading-anchor" href="#Computing-energies-and-loglikelihoods">Computing energies and loglikelihoods</a><a id="Computing-energies-and-loglikelihoods-1"></a><a class="docs-heading-anchor-permalink" href="#Computing-energies-and-loglikelihoods" title="Permalink"></a></h3><p>Given a sequence, the package allows to compute easily its energy assigned by a model</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; model = DimerForce(example_seq, [&quot;CG&quot;]);</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DinucleotideForces.ComputeEnergy(example_seq, model)</code><code class="nohighlight hljs ansi" style="display:block;">-502.5299370908549</code></pre><p>and its log-likelihood</p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DinucleotideForces.ComputeLoglikelihood(example_seq, model)</code><code class="nohighlight hljs ansi" style="display:block;">-489.14789197001</code></pre><p>To make sense of this large negative log-likelihood, consider that in the uniform case we have <span>$4^L$</span> sequences, so the log-likelihood of each of them would be:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; - length(example_seq) * log(4)</code><code class="nohighlight hljs ansi" style="display:block;">-507.38373616987997</code></pre><p>Notice that the function <code>ComputeLoglikelihood</code> computes the partition function each time it is runned. This can slow down the computation when the function is called many times with the same model and to prevent this the log of the partition function can be passed directly:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; model = DimerForce(example_seq, [&quot;CG&quot;, &quot;AC&quot;]);</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; ks = keys(model);</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; k1 = [k for k in ks if length(k)==1];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; k2 = [k for k in ks if length(k)==2];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; fields = [model[k] for k in k1];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; forces = [model[k] for k in k2];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; logZ = DinucleotideForces.eval_log_Z(fields, forces, k2, length(example_seq));</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: UndefVarError: eval_log_Z not defined</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DinucleotideForces.ComputeLoglikelihood(example_seq, model; logZ=logZ)</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: UndefVarError: logZ not defined</code></pre><h3 id="Sampling-sequences"><a class="docs-heading-anchor" href="#Sampling-sequences">Sampling sequences</a><a id="Sampling-sequences-1"></a><a class="docs-heading-anchor-permalink" href="#Sampling-sequences" title="Permalink"></a></h3><p>The framework used by this package introduced a probability distribution over the sequences given a set of fields and/or forces. Sequences can be sampled from this distribution exactly (that is, without using any Markov-chain based algorithm such as Metropolis) using the simple topology of the graph of interactions between nucleotides (a tree) and the fact that the partition function can be computed exactly. The funcion <code>SampleSequence</code> of the package implements this idea, and it is very easy to use, its only arguments being the model that has to be used to sample from and the length of the desired sequence:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; nts = [&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [a*b for a in nts for b in nts];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; model = DimerForce(example_seq, motifs);</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DinucleotideForces.SampleSequence(model, 1000)</code><code class="nohighlight hljs ansi" style="display:block;">&quot;GACGATTAGTTCCATTACTGAAGGATAAATGAACAAATAAGAGAGAGGCACCAGACGCCGATACTCCTGTGAGGGGCTACCGAAAACCACACGCAAACCAGTACACGAGCTTTGCCATAAAGCTGGACGAATCCCAAGATAAACAACAGGGTGCCATGGGGACAGAAATCAAGAAACTTAGTGAGCGGACCAAAGCTCAAAGGCACAAATCAACTAGTGGGAGAGTAAAGGGCCCTTTAAAATTAAGAATGTTCACCGCGGCCAGATGAAA&quot; ⋯ 459 bytes ⋯ &quot;CCTCCAAACCCATGCCCCAGCCATGAGGGGGCCCCATAGGAAAACGGGGTCACACCCTGAATAAATTCAGGTCAGATAGAAAAATTACAGAGAGGGGCGGAGATTAACGGAGAGGACTCAAGTGACGAGATGCATAATGGATGACGCTGAGGGACATCTGCACGCATATGTGCTTCTGGAAAATTCCCCCAAGAAGATTGCAGAGTTAGGATCAACTCTTGCCAATTCCACTACGGGGGAGGATTTGTACCCACCAGGTGTGAGGGAGGC&quot;</code></pre><h2 id="Known-issues"><a class="docs-heading-anchor" href="#Known-issues">Known issues</a><a id="Known-issues-1"></a><a class="docs-heading-anchor-permalink" href="#Known-issues" title="Permalink"></a></h2><p>There are issues with the <code>add_pseudocount</code> argument of <code>DimerForce</code>.  In principle, using <code>add_pseudocount=true</code> should add a single pseudocount to the number of each motif inferred (nucleotide or di-nucleotide). However, from the output of the following two examples it looks that something is not working properly</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; nts = [&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; motifs = [a*b for a in nts for b in nts];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 20 entries:
  &quot;T&quot;  =&gt; -1.76995
  &quot;C&quot;  =&gt; -1.53967
  &quot;CC&quot; =&gt; 0.0
  &quot;GC&quot; =&gt; 0.203128
  &quot;GG&quot; =&gt; 0.0
  &quot;CG&quot; =&gt; -1.25295
  &quot;AT&quot; =&gt; 0.0
  &quot;A&quot;  =&gt; -1.04555
  &quot;CA&quot; =&gt; 0.206446
  &quot;TG&quot; =&gt; 0.0859823
  &quot;TA&quot; =&gt; -0.144103
  &quot;GT&quot; =&gt; 0.0
  &quot;G&quot;  =&gt; -1.33292
  &quot;GA&quot; =&gt; 0.790941
  &quot;TT&quot; =&gt; 0.0
  &quot;AC&quot; =&gt; -0.185393
  &quot;CT&quot; =&gt; 0.0
  &quot;AA&quot; =&gt; 0.0
  &quot;AG&quot; =&gt; -0.639832
  &quot;TC&quot; =&gt; 0.403223</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(example_seq, motifs, add_pseudocount=true)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 20 entries:
  &quot;T&quot;  =&gt; NaN
  &quot;C&quot;  =&gt; NaN
  &quot;CC&quot; =&gt; NaN
  &quot;GC&quot; =&gt; NaN
  &quot;GG&quot; =&gt; NaN
  &quot;CG&quot; =&gt; NaN
  &quot;AT&quot; =&gt; NaN
  &quot;A&quot;  =&gt; NaN
  &quot;CA&quot; =&gt; NaN
  &quot;TG&quot; =&gt; NaN
  &quot;TA&quot; =&gt; NaN
  &quot;GT&quot; =&gt; NaN
  &quot;G&quot;  =&gt; NaN
  &quot;GA&quot; =&gt; NaN
  &quot;TT&quot; =&gt; 0.0
  &quot;AC&quot; =&gt; NaN
  &quot;CT&quot; =&gt; NaN
  &quot;AA&quot; =&gt; NaN
  &quot;AG&quot; =&gt; NaN
  &quot;TC&quot; =&gt; NaN</code></pre><p>This is even worse if a shorter sequence is used</p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; ex2 = example_seq[1:100]</code><code class="nohighlight hljs ansi" style="display:block;">&quot;ATGGAGAGAATAAAAGAATTAAGGGATCTAATGTCACAGTCCCGCACTCGCGAGATACTAACAAAAACCACTGTGGACCATATGGCCATAATCAAGAAGT&quot;</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(ex2, motifs)</code><code class="nohighlight hljs ansi" style="display:block;">Dict{String, Float64} with 20 entries:
  &quot;T&quot;  =&gt; -1.88382
  &quot;C&quot;  =&gt; -1.41788
  &quot;CC&quot; =&gt; 0.0
  &quot;GC&quot; =&gt; -0.513471
  &quot;GG&quot; =&gt; 0.0
  &quot;CG&quot; =&gt; -0.508183
  &quot;AT&quot; =&gt; 0.0
  &quot;A&quot;  =&gt; -0.980651
  &quot;CA&quot; =&gt; 0.147666
  &quot;TG&quot; =&gt; 0.250867
  &quot;TA&quot; =&gt; 0.262363
  &quot;GT&quot; =&gt; 0.0
  &quot;G&quot;  =&gt; -1.46664
  &quot;GA&quot; =&gt; 0.262805
  &quot;TT&quot; =&gt; 0.0
  &quot;AC&quot; =&gt; -0.439822
  &quot;CT&quot; =&gt; 0.0
  &quot;AA&quot; =&gt; 0.0
  &quot;AG&quot; =&gt; -0.303644
  &quot;TC&quot; =&gt; 0.248225</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DimerForce(ex2, motifs, add_pseudocount=true)</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: LinearAlgebra.SingularException(1)</code></pre><h2 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h2><ul><li><a href="#DinucleotideForces.ComputeEnergy-Tuple{AbstractString, Dict{String, Float64}}"><code>DinucleotideForces.ComputeEnergy</code></a></li><li><a href="#DinucleotideForces.ComputeLoglikelihood-Tuple{AbstractString, Dict{String, Float64}}"><code>DinucleotideForces.ComputeLoglikelihood</code></a></li><li><a href="#DinucleotideForces.ComputeNumObservations-Tuple{Vector{&lt;:AbstractString}, Union{Vector{&lt;:AbstractString}, Vector{&lt;:AbstractChar}}, Int64}"><code>DinucleotideForces.ComputeNumObservations</code></a></li><li><a href="#DinucleotideForces.DimerForce"><code>DinucleotideForces.DimerForce</code></a></li><li><a href="#DinucleotideForces.DimerForceOnlyForces"><code>DinucleotideForces.DimerForceOnlyForces</code></a></li><li><a href="#DinucleotideForces.DimerForceWithFields"><code>DinucleotideForces.DimerForceWithFields</code></a></li><li><a href="#DinucleotideForces.EvaluateLogZ-NTuple{4, Any}"><code>DinucleotideForces.EvaluateLogZ</code></a></li><li><a href="#DinucleotideForces.FixFinalGauge!-Tuple{Dict{String, Float64}}"><code>DinucleotideForces.FixFinalGauge!</code></a></li><li><a href="#DinucleotideForces.GaugeAwayVariables-Tuple{Vector{&lt;:AbstractString}, Bool}"><code>DinucleotideForces.GaugeAwayVariables</code></a></li><li><a href="#DinucleotideForces.GenerateMotifMatrix-Tuple{Vector{&lt;:AbstractString}}"><code>DinucleotideForces.GenerateMotifMatrix</code></a></li><li><a href="#DinucleotideForces.GenerateNucleotideMatrix-Tuple{Vector{Float64}, Bool}"><code>DinucleotideForces.GenerateNucleotideMatrix</code></a></li><li><a href="#DinucleotideForces.Marginal1GivenPrevious-Tuple{Dict{String, Float64}, AbstractString}"><code>DinucleotideForces.Marginal1GivenPrevious</code></a></li><li><a href="#DinucleotideForces.Marginal1Point-Tuple{Dict{String, Float64}}"><code>DinucleotideForces.Marginal1Point</code></a></li><li><a href="#DinucleotideForces.Marginal2Points-Tuple{Dict{String, Float64}, Int64, Int64}"><code>DinucleotideForces.Marginal2Points</code></a></li><li><a href="#DinucleotideForces.PreprocessSequences-Tuple{Vector{&lt;:AbstractString}}"><code>DinucleotideForces.PreprocessSequences</code></a></li><li><a href="#DinucleotideForces.SampleSequence"><code>DinucleotideForces.SampleSequence</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.ComputeEnergy-Tuple{AbstractString, Dict{String, Float64}}" href="#DinucleotideForces.ComputeEnergy-Tuple{AbstractString, Dict{String, Float64}}"><code>DinucleotideForces.ComputeEnergy</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputeEnergy(seq::AbstractString, fields_forces::Dict{String, Float64})</code></pre><p>Given a sequence seq, the single-nucleotide fields and the dinucleotide forces (as a single dict),  compute the energy of this sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L371-L376">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.ComputeLoglikelihood-Tuple{AbstractString, Dict{String, Float64}}" href="#DinucleotideForces.ComputeLoglikelihood-Tuple{AbstractString, Dict{String, Float64}}"><code>DinucleotideForces.ComputeLoglikelihood</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputeLoglikelihood(seq::AbstractString, fields_forces::Dict{String, Float64}; logZ=missing)</code></pre><p>Given a sequence seq, the single-nucleotide fields and the dinucleotide forces (as a single dict), compute the log-likelihood (energy minus log of Z) of this sequence. logZ can be passed directly if pre-computed, otherwise is it computed each time this function is called.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L386-L392">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.ComputeNumObservations-Tuple{Vector{&lt;:AbstractString}, Union{Vector{&lt;:AbstractString}, Vector{&lt;:AbstractChar}}, Int64}" href="#DinucleotideForces.ComputeNumObservations-Tuple{Vector{&lt;:AbstractString}, Union{Vector{&lt;:AbstractString}, Vector{&lt;:AbstractChar}}, Int64}"><code>DinucleotideForces.ComputeNumObservations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputeNumObservations(seqs::Vector{&lt;:AbstractString}, independent_motifs::Vector{&lt;:AbstractString}, L::Int)</code></pre><p>For each motif in <code>independent_motifs</code>, compute the number of observed motifs in each sequence in <code>seqs</code>, then divide by the sequence length, take the average of these intensive fractions over the sequences, and multiply by the model length L.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L191-L196">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.DimerForce" href="#DinucleotideForces.DimerForce"><code>DinucleotideForces.DimerForce</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">DimerForce(seq::Union{AbstractString,Vector{&lt;:AbstractString}}, motifs::Vector{&lt;:AbstractString}; L::Union{Int,Missing}=missing, 
                freqs::Union{Vector{Float64},Missing}=missing, tolerance::Float64=0.01, max_iter::Int=100, 
                add_pseudocount=false)</code></pre><p>If frequencies are given, return the forces on the motifs &#39;motifs&#39; computed for sequence &#39;seq&#39; with  the frequency bias given. If frequencies are not given, the local fields (NOT frequencies!) are inferred and returned (in a gauge such that the exp of them sum to 1, so the fields can be interpreted as log of frequencies). Notice that seq is used only to compute the number of observed motifs (and nucleotides if the frequencies are not provided by user).  Also notice that if motifs are dinucleotides, only up to 12 of them are  independent.  Finally, tolerance and max<em>iter are parameters for the Newton-Raphson algorithm  used to solve the system of equations. If add</em>pseudocount, a single pseudocount is added for each observed number of nucleotides and dinucleotides. seq can be a single sequence or a vector of sequences. L can be used to define the length of the modeled sequence. If not given, the length of the sequence will be used if a single  sequence or a set of sequences of the same lengths are provided. Otherwise, the default value of 5000 will be used.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L321-L340">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.DimerForceOnlyForces" href="#DinucleotideForces.DimerForceOnlyForces"><code>DinucleotideForces.DimerForceOnlyForces</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">DimerForceOnlyForces(seqs::Vector{&lt;:AbstractString}, motifs::Vector{&lt;:AbstractString}, fields::Vector{Float64},
                            Lmodel::Union{Int,Missing}=missing; tolerance::Float64=0.01, max_iter::Int=100, add_pseudocount=true)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L276-L280">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.DimerForceWithFields" href="#DinucleotideForces.DimerForceWithFields"><code>DinucleotideForces.DimerForceWithFields</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">DimerForceWithFields(seqs::Vector{&lt;:AbstractString}, nucleotides::Vector{Char}, motifs::Vector{&lt;:AbstractString}, 
                 Lmodel::Union{Int,Missing}=missing; tolerance::Float64=0.01, max_iter::Int=100, add_pseudocount=true)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L203-L207">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.EvaluateLogZ-NTuple{4, Any}" href="#DinucleotideForces.EvaluateLogZ-NTuple{4, Any}"><code>DinucleotideForces.EvaluateLogZ</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Compute the transfer matrix through GenerateMotifMatrix, generate_freqsM and the forces given, then takes the correct power to obtian Z.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L41-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.FixFinalGauge!-Tuple{Dict{String, Float64}}" href="#DinucleotideForces.FixFinalGauge!-Tuple{Dict{String, Float64}}"><code>DinucleotideForces.FixFinalGauge!</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Returns a new dictionary with all the fields/couplings inferred, such that the exp of the fields sum to 1 and as much information as possible is included in the field, by putting other motifs to 0 (if possible, those of the form AA, CC, GG, TT and those ending  with T).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L126-L132">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.GaugeAwayVariables-Tuple{Vector{&lt;:AbstractString}, Bool}" href="#DinucleotideForces.GaugeAwayVariables-Tuple{Vector{&lt;:AbstractString}, Bool}"><code>DinucleotideForces.GaugeAwayVariables</code></a> — <span class="docstring-category">Method</span></header><section><div><p>This function returns the nucleotides or motifs that have to be inferred after as many variables as possible are fixed to 0 through gauge transformations.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L69-L72">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.GenerateMotifMatrix-Tuple{Vector{&lt;:AbstractString}}" href="#DinucleotideForces.GenerateMotifMatrix-Tuple{Vector{&lt;:AbstractString}}"><code>DinucleotideForces.GenerateMotifMatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Return the matrices used for the computation of the partition function for each motif in motifs. In particular, M_{ij} is the number of times  the given motif appears in n(i) + n(j), with n(i) the i-th nucleotide.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L16-L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.GenerateNucleotideMatrix-Tuple{Vector{Float64}, Bool}" href="#DinucleotideForces.GenerateNucleotideMatrix-Tuple{Vector{Float64}, Bool}"><code>DinucleotideForces.GenerateNucleotideMatrix</code></a> — <span class="docstring-category">Method</span></header><section><div><p>Return a matrix with the informations about the nt frequences, so that when it is multiplied elementwise with that obtained through  GenerateMotifMatrix (after implementing the exponentiation with the force, see EvaluateLogZ), the transfer matrices used to compute Z are finally obtained.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L26-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.Marginal1GivenPrevious-Tuple{Dict{String, Float64}, AbstractString}" href="#DinucleotideForces.Marginal1GivenPrevious-Tuple{Dict{String, Float64}, AbstractString}"><code>DinucleotideForces.Marginal1GivenPrevious</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Marginal1GivenPrevious(Marginal2Points::Dict{String, Float64}, fixed::AbstractString; first::Bool=true)</code></pre><p>Given the 2point marginal, fix the first symbol (if <code>first</code>=true, otherwise  fix the second) to obtain the probability of the second given that the first is equal to <code>fixed</code> (or viceversa if <code>first</code>=false).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L546-L552">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.Marginal1Point-Tuple{Dict{String, Float64}}" href="#DinucleotideForces.Marginal1Point-Tuple{Dict{String, Float64}}"><code>DinucleotideForces.Marginal1Point</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Marginal1Point(Marginal2Points::Dict{String, Float64}; first::Bool=true)</code></pre><p>Given the 2point marginal, sum over the second symbol (if first=true, otherwise  over the first) to obtain the 1-point marginal.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L524-L529">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.Marginal2Points-Tuple{Dict{String, Float64}, Int64, Int64}" href="#DinucleotideForces.Marginal2Points-Tuple{Dict{String, Float64}, Int64, Int64}"><code>DinucleotideForces.Marginal2Points</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Marginal2Points(fields_forces::Dict{String, Float64}, L::Int, pos::Int)</code></pre><p>Given a model specified by fields and forces and sequence length, compute the 2-point marginal in position (i-1, i). The result is returned as a dictionary &quot;s<em>(i-1) s</em>i&quot; =&gt; probability.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L410-L415">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.PreprocessSequences-Tuple{Vector{&lt;:AbstractString}}" href="#DinucleotideForces.PreprocessSequences-Tuple{Vector{&lt;:AbstractString}}"><code>DinucleotideForces.PreprocessSequences</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">PreprocessSequences(seqs_in::Vector{&lt;:AbstractString})</code></pre><p>Pre-process the vector of sequences <code>seqs_in</code>, so that sequences only contains letters A, C, G, T. The pre-processed sequence vector is the returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L173-L177">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="DinucleotideForces.SampleSequence" href="#DinucleotideForces.SampleSequence"><code>DinucleotideForces.SampleSequence</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">SampleSequence(fields_forces::Dict{String, Float64}, L::Int, 
                    rng::AbstractRNG=Xoshiro(123))</code></pre><p>Sample a sequence of length <code>L</code> from the model specified by the parameters in <code>fields_forces</code>. The sampling is exact (that is  the correct probability distribution is sampled) and a random  generator can be specified via <code>rng</code>. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/DinucleotideForces.jl/blob/04cec2d527371b39a499767ac94a1da9e1f85125/src/DinucleotideForces.jl#L575-L583">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Tuesday 21 February 2023 16:38">Tuesday 21 February 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
